feat: pluginMinifyHtmlLiterals #1568

vegarringdal opened the pull request:
@nchanged @kyr0 

Have tried implement #1568.
Hopefully I havent messed up to much :joy:

nchanged commented on src/plugins/core/plugin_minify_html_literals.ts:
```diff
@@ -0,0 +1,33 @@
+  return (ctx: Context) => {
+    ctx.ict.on('assemble_fast_analysis', props => {
+      const { module } = props;
+      if (module.props && (module.props.extension === '.ts' || module.props.extension === '.js')) {
```
`module.isExecutable()` ;-)

vegarringdal replied:
Very nice, will update

nchanged commented on src/plugins/core/plugin_minify_html_literals.ts:
```diff
@@ -0,0 +1,33 @@
+    ctx.ict.on('assemble_fast_analysis', props => {
+      const { module } = props;
+      if (module.props && (module.props.extension === '.ts' || module.props.extension === '.js')) {
+        if (module.contents.indexOf('html`') !== -1 || module.contents.indexOf('css`') !== -1) {
```
we can remove that probably

vegarringdal replied:
I figured it was best to check first. If there is not template tags with those names we dont need to run the plugin, indexof will be faster then using the plugin. 
Want to keep fusebox fast :-)

nchanged commented on src/plugins/core/plugin_minify_html_literals.ts:
```diff
@@ -0,0 +1,33 @@
+import { minifyHTMLLiterals } from 'minify-html-literals';
+import { Context } from '../../core/Context';
+
+export function pluginMinifyHtmlLiterals(...options: any) {
```
can you add an interface IPluginMinifyHtmlLiterals to follow the conventions?

vegarringdal replied:
yes

nchanged commented on src/plugins/core/plugin_minify_html_literals.ts:
```diff
@@ -0,0 +1,33 @@
+import { minifyHTMLLiterals } from 'minify-html-literals';
```
don't import it here ;_)
do it like this: 
https://github.com/fuse-box/fuse-box/blob/master/src/plugins/core/plugin_less.ts#L11-L17

cuz that package is in dev dependencies it will simply fail

vegarringdal replied:
ok.
just do `require('minify-html-literals')` before I use it then ?

nchanged commented on src/plugins/core/plugin_minify_html_literals.ts:
```diff
@@ -0,0 +1,33 @@
+export function pluginMinifyHtmlLiterals(...options: any) {
+  const useroptions = options;
+  return (ctx: Context) => {
+    ctx.ict.on('assemble_fast_analysis', props => {
```
I also think it should be at `bundle_resolve_module` level, `assemble_fast_analysis` is a bit too early

vegarringdal replied:
ok, I really dont know what comes first and last atm.
So will just update like you say :-)

vegarringdal commented:
Implemented the review comments.
keept the check if html/css template strings only. will be faster I think.

![image](https://user-images.githubusercontent.com/2901416/64477221-168c3680-d199-11e9-8fc9-d40cca1124a2.png)

added quickfix for thsi for now with any.
`cssBundle.override((response as any).styles, (response as any).sourceMap.toString());`

vegarringdal commented:
@nchanged added what we talked about. have a last look :-)

vegarringdal commented on src/plugins/core/plugin_minify_html_literals.ts:
```diff
@@ -0,0 +1,53 @@
+          try {
+            const minifyHTMLLiterals = require('minify-html-literals');
+            newcontent = minifyHTMLLiterals(module.contents, ...useroptions);
+          } catch {}
```
Guessing you saw I did a error if it did not generate code since its merged :-)

nchanged replied:
yeah but still it's a bit weird tho ;-)
